### PR TITLE
[ZEPPELIN-707]Automatically adds %.* of previous paragraph's typing

### PR DESCRIPTION
**Pull request by mwkang**
### What is this PR for?
Automatically adds %pyspark in the code area after paragraph created, if previous paragraph's typing is %pyspark.

### What type of PR is it?
New Feature

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-707

### How should this be tested?
unit test
Run-time checking

### Screenshots (if appropriate)
- default interpreter
zeppelin-707
- while run paragraph, go to line end
zeppelin-707-2

### CODE REVIEW

**Comment by Leemoonsoo** [comment]
Zeppelin automatically adds new paragraph when last paragraph runs.
Isn't it make sense to add %.* to this auto added paragraph?

**Comment by mwkang** [comment]
I agree with you. This commit is too hasty.
I'm so sorry. I'm looking in that.

**Comment by mwkang** [comment]
@Leemoonsoo 
I change some code. Could you mind if I request to review this?

**Comment by corneadoug** [comment]
Isn't it possible not to have any front-end code for that?
Why not just include the `%..` in the empty note in the backend?

**Comment by mwkang** [comment]
The purpose of this front-end code is to track the recent usage of the interpteter(by the users).
Okay. include the `%...` in the empty note.

**Comment by felixcheung** [comment]
Does this work when the user has not included `%interpreter` in the previous paragraph? (ie. using the default interpreter)

**Comment by mwkang** [comment]
@felixcheung 
Yes, it works.
If user doesn't any %interpreter in previous paragraphs, `%..`(default interpreter) will be added.
If user add %interpreter in previous paragraphs, %interpreter will be added.

e.g.)
paragraph-1 (no interpreters) -> paragraph-2 (no interpreters) -> paragraph-3 (`%..` will be added.)
paragraph-1 (%sh) -> paragraph-2 (no interpreters) -> paragraph-3 (`%sh` will be added.)

**Comment by mwkang** [comment]
Is there anything wrong?

**Comment by bzz** [comment]
@mwkang CI is failing now, you probably need to check in `zeppelin-webapp` what does `./grunt jshint` complain about.

**Comment by corneadoug** [comment]
For me, same comment as before.
There is no need for any front-end code.

**Comment by mwkang** [comment]
@bzz Thank you for your review.
@corneadoug Okay, I will change the code does not use a front-end code.

**Comment by mwkang** [comment]
@corneadoug I have a question. If User doesn't run paragraph, Always add `%..`. Because I couldn't know what is recent usage of the interpreter(by the users). Is it okay?

**Comment by corneadoug** [comment]
If we create a new empty paragraph, we should add the `%..` (default, or previous).
So this should be working for: running the last paragraph of the note + adding a new paragraph

**Comment by mwkang** [comment]
I change my code it is use only backend. I force push my remote branch. So remove previous commit.

**Comment by mwkang** [comment]
@corneadoug 
Could you review this?

**Comment by corneadoug** [comment]
@mwkang 

I found 2 problems:

1) Content of the first line is copied instead of interpreter sign only, for example:
```
%md hello
```
Will copy  `%md hello` to the new paragraph, instead of `%md `

2) Default interpreter is not copied
If you type code using the default interpreter, it will add the previous paragraph interpreter sign instead.
So If you have:

#### Paragraph 1
```
%md
hello
```

#### Paragraph 2
```
println("Hello")
```

The new paragraph will have `%md`

**Comment by corneadoug** [comment]
@mwkang Do you have time to fix this PR?

**Comment by mwkang** [comment]
@corneadoug 
Yes, Sorry for delay.

**Comment by mwkang** [comment]
@corneadoug 
I fixed that problems.
Could you mind if I request review to you?

**Comment by mwkang** [comment]
I'm so sorry. CI was failed.
I'll try to fix it.

**Comment by corneadoug** [comment]
@mwkang no problem
It seems some tests and selectors are broken with the new changes.
I will try the branch again and post if I find what is breaking at the same time

**Comment by mwkang** [comment]
@corneadoug @Leemoonsoo Yeah! I passed all test.

**Comment by corneadoug** [comment]
@mwkang 
Sorry for late answer, I tested again and have 2 comments:
* When you create a new notebook, the interpreter is listed as: `%..`. Any way to get the default interpreter from back-end and insert it instead?
* After running a query, the cursor is placed on the next paragraph where the interpreter name was injected, however the cursor is placed at the beginning of the line (so you can't start typing)

**Comment by mwkang** [comment]
@corneadoug 
Thank you for your comment.
I am sorry I don't understand clearly.
Could you explain more?
What is the meaning get the default interpreter from back-end and insert it instead?

**Comment by corneadoug** [comment]
I don't remember if there was a way to change the default interpreter.
But at least without config, it uses '%spark' by default, so
`sc.version` == `%spark sc.version`

In our case, if we try running
```
%..
sc.version
```

You will get a: `%.. interpreter not found`

**Comment by mwkang** [comment]
@corneadoug 
Oh... I don't consider about that.
Thanks for mention that,
I will fix my code.

**Comment by mwkang** [comment]
- [ ] When you create a new notebook, the interpreter is listed as: %... Any way to get the default interpreter from back-end and insert it instead?
- [ ] After running a query, the cursor is placed on the next paragraph where the interpreter name was injected, however the cursor is placed at the beginning of the line (so you can't start typing)

**Comment by mwkang** [comment]
@corneadoug 
Hello, I have a question.
Only after running a query, the cursor go to line end in new paragraph.
If It is added manually, the cursor doesn't go to line end.
Added manually meaning is using paragraph add menu or click add button between two paragraphs.
Is it Okay?
Or Have I to make the cursor go to line end without exception?

**Comment by corneadoug** [comment]
Since the cursor is moving in the text editor of the new paragraph, it should also go at the end of the line

**Comment by mwkang** [comment]
@corneadoug I fix whole issue. Can you check my code?

**Comment by corneadoug** [comment]
@mwkang I still get `%..` when I create a new note, is it normal?

Could you also add a space after the %interpreter when you add it to the note?
That way we can start typing directly without having to add that space

@jongyoul If you could test that using some jdbc or other interpreters cases, that would be nice :)

**Comment by jongyoul** [comment]
It works well with another default interpreter except `%spark`. See the below.
screen shot 2016-06-10 at 6 05 11 pm
screen shot 2016-06-10 at 6 05 22 pm

But I also don't think it's good idea to display `%..` by default. It would be better display actual default name of interpreter like `%spark` or `%jdbc`. It prevent new users from confusing and trying to understand what that means.

@mwkang Could you please try to fix it? and OccupiedInterpreter looks redundant because it stores noteId-interpreterName map only and returns interpreterName with noteId. This can be parsed from InterpreterFactory or another class directly. What's the benefit to create OccupiedInterpreter class?

I also think it's simple to choose default interpreter when we create notebook and store it into note.json and use it.

**Comment by mwkang** [comment]
Okay. I am going to fix it.
I show real interpreter name instead of '%..'

@jongyoul I think about single responsibility. But if you have thought it is redundant, I will change code.
You think it is redundant, don't you?

(It's redundant.. And I think I lack of develop experience. So I sometimes have a mistake. But I have a can-do attitdue to learn from other's opinion.)

**Comment by mwkang** [comment]
I'll fix after this branch rebases onto develop.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/NoteInterpreterLoader.java [hunk 1]
```diff
@@ -192,4 +192,16 @@ public Interpreter get(String replName) {
...
+    return settings.get(0);
+  }
+
+  Optional<InterpreterSetting> getDefaultInterpreterSetting() {
+    List<InterpreterSetting> settings = getInterpreterSettings();
+    if (settings == null || settings.isEmpty()) {
+      return Optional.absent();
+    }
+    return Optional.of(settings.get(0));
+  }
```
Comment (jongyoul) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/NoteInterpreterLoader.java]: Why do you make this method to return optional, not just returning getDefaultInterpreterSetting(getInterpreterSettings())? It looks like that Note.getDefaultInterpreterName uses this method only. If you want to check nullable list and emptiness, should you check the above method of this method?

Comment (mwkang) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/NoteInterpreterLoader.java]: Oh.. thank you for your comment.
I will change my code.

**Comment by mwkang** [comment]
@jongyoul @corneadoug 
Can you review my codes?
(I had rebased my branch because branch had conflict)

(Um.. CI is failed. :cry: )

**Comment by jongyoul** [comment]
@mwkang Could you please re-trigger this PR? Close and reopen this issue.

**Comment by jongyoul** [comment]
There's no test case for your implementation. Could you please add some test cases on your feature? I think NoteTest is one of the best point to write tests

**Comment by mwkang** [comment]
I am not sure that this test case is you expected

**Comment by jongyoul** [comment]
@mwkang Those tests are enough. I've tested it and LGTM. Thanks for contribution.

**Comment by jongyoul** [comment]
@mwkang I have one more question. See below.
screen shot 2016-06-17 at 11 03 44 pm

Can you fix the case of wrong interpreter name? Or do you have any idea to deal with?

**Comment by mwkang** [comment]
@jongyoul If user enter wrong name, I ignore that name. Is it okay? 
Or do I try to figure similar name out ?

**Comment by jongyoul** [comment]
@mwkang It would be ok to ignore it. BTW, I suggest that you handle this issue with another PR because you should rebase your PR again and again. If you rebase this PR this time, I'll merge it into master.

**Comment by mwkang** [comment]
@jongyoul I rebase this PR onto master. I'll create new PR about It ignore wrong interpreter name

**Comment by jongyoul** [comment]
@mwkang Could you please re-trigger test once again?

**Comment by mwkang** [comment]
I reopened this PR. But I think I have to fix some test case.

**Comment by mwkang** [comment]
- [ ] https://travis-ci.org/apache/zeppelin/jobs/138874097
`ZeppelinSparkClusterTest.sparkRTest:90`

- [ ] https://travis-ci.org/apache/zeppelin/jobs/138874103
```
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 104.235 sec - in org.apache.zeppelin.integration.ZeppelinIT

Results :

Failed tests: 
  ParagraphActionsIT.testTitleButton:343 After Show Title : The title field contains
Expected: "Untitled"
     but: was ""

Tests in error: 
  ParagraphActionsIT.testRemoveButton:143->AbstractZeppelinIT.waitForParagraph:84->AbstractZeppelinIT.pollingWait:110 쨩 Timeout
  ParagraphActionsIT.testMoveUpAndDown:180->AbstractZeppelinIT.waitForParagraph:84->AbstractZeppelinIT.pollingWait:110 쨩 Timeout
  ParagraphActionsIT.testCreateNewButton:75->AbstractZeppelinIT.waitForParagraph:84->AbstractZeppelinIT.pollingWait:110 쨩 Timeout
```

**Comment by mwkang** [comment]
Wow thanks for your patience. This PR is a long period and lots of comments. I'll keep on trying my best.

**Pull request closed without merging**